app/controllers/exercise_solutions_controller.rb

```diff
@@ -4,9 +4,16 @@ class ExerciseSolutionsController < ApplicationController
   before_action :authenticate!
 
   def create
+    guide_previously_done = is_guide_done?
     @solution = @exercise.submit_solution(current_user, solution_params)
     @solution.run_tests!
-    render partial: 'exercise_solutions/results', locals: {solution: @solution}
+    guide_finished_by_solution = !guide_previously_done && is_guide_done?
+    render partial: 'exercise_solutions/results', locals: {solution: @solution, guide_finished_by_solution: guide_finished_by_solution}
+  end
+
+  def is_guide_done?
+    guide = @exercise.guide
+    guide && guide.stats_for(current_user).done?
   end
```

Comment:
Also, it could be a guide instance method. That way we could test it. 

I wonder whether it already exists....